### PR TITLE
Copy OPD data when initializing FITSOpticalElement with HDUList

### LINKED ISSUE #89: POPPY FITSOpticalElement modifies OPD arrays in-place in surprising ways
It's possible to assign an HDUList to Instrument.pupilopd instead of a path to a FITS file. However, FITSOpticalElement modifies the OPD data in-place to apply a unit normalization.

This means that performing calcPSF on an Instrument more than once with the same HDUList assigned to Instrument.pupilopd (and thus, instantiating a new FITSOpticalElement more than once with the same HDUList) destructively modifies the underlying OPD array.

To illustrate, suppose you have loaded an OPD map for use with webbpsf.NIRCam (a poppy.Instrument subclass). Using `np.ndarray.setflags` with `write=False` means any attempt to modify the OPD data in place will raise an exception.

```python
from webbpsf import NIRCam
from astropy.io import fits
opdmap = fits.open('./OPD_RevV_nircam_132.fits')
opdmap[0].data.setflags(write=False)  # mark opd data as read-only
nc = NIRCam()
nc.pupilopd = opdmap
nc.calcPSF()
```

```
INFO:webbpsf:Setting up PSF calculation for NIRCam
WARNING:webbpsf:Filter F200W not found in lookup table for default number of wavelengths to use.. setting default nlambda=10
WARNING:poppy:Pysynphot unavailable (or invalid source supplied)!   Assuming flat # of counts versus wavelength.
WARNING:poppy:CAUTION: no WAVEUNIT keyword found in filter file /Users/jdl/dev/webbpsf-data/NIRCam/filters/F200W_throughput.fits. Assuming = Angstroms by default
WARNING:poppy:CAUTION: Just interpolating rather than integrating filter profile, over 10 steps
INFO:webbpsf:NIRCam pixel scale updated to 0.031700 arcsec/pixel to match the requested wavelength range.
INFO:webbpsf:PSF calc using fov_arcsec = 5.000000, oversample = 4, number of wavelengths = 10
INFO:webbpsf:Creating optical system model:
INFO:poppy:Initialized OpticalSystem: JWST+NIRCam
INFO:poppy:JWST Pupil: Loaded amplitude transmission from /Users/jdl/dev/webbpsf-data/pupil_RevV.fits
INFO:poppy:JWST Pupil: Loaded OPD from supplied fits.HDUList object
INFO:poppy:The supplied pupil OPD is a datacube but no slice was specified. Defaulting to use slice 0.
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-c7edfe5da60e> in <module>()
      5 nc = NIRCam()
      6 nc.pupilopd = opdmap
----> 7 nc.calcPSF()

/Users/jdl/dev/webbpsf/webbpsf/webbpsf_core.pyc in calcPSF(self, outfile, source, filter, nlambda, monochromatic, fov_arcsec, fov_pixels, oversample, detector_oversample, fft_oversample, calc_oversample, rebin, clobber, display, return_intermediates, **kwargs)
    407         #  instantiate an optical system using the current parameters
    408         self.optsys = self._getOpticalSystem(fov_arcsec=fov_arcsec, fov_pixels=fov_pixels,
--> 409             fft_oversample=fft_oversample, detector_oversample=detector_oversample, options=local_options)
    410         # and use it to compute the PSF (the real work happens here, in code in poppy.py)
    411         #result = self.optsys.calcPSF(source, display_intermediates=display, save_intermediates=save_intermediates, display=display)

/Users/jdl/dev/webbpsf/webbpsf/webbpsf_core.pyc in _getOpticalSystem(self, fft_oversample, detector_oversample, fov_arcsec, fov_pixels, options)
    561                 opd=opd_map,
    562                 opdunits='micron',
--> 563                 rotation=self._rotation
    564             )
    565         self.pupil_radius = pupil_optic.pupil_diam / 2.0

/Users/jdl/dev/poppy/poppy/poppy_core.py in addPupil(self, optic, function, **kwargs)
    960         elif optic is None and len(kwargs) > 0: # create image from files specified in kwargs
    961             # create image from files specified in kwargs
--> 962             optic = FITSOpticalElement(planetype=_PUPIL, oversample=self.oversample, **kwargs)
    963         elif optic is None and len(kwargs) == 0: # create empty optic.
    964             from .import optics

/Users/jdl/dev/poppy/poppy/poppy_core.py in __init__(self, name, transmission, opd, opdunits, shift, rotation, pixelscale, planetype, transmission_index, opd_index, **kwargs)
   2088                 pass # no need to rescale
   2089             elif opdunits.lower() == 'micron' or opdunits.lower() == 'um' or opdunits.lower() == 'micrometer':
-> 2090                 self.opd *= 1e-6
   2091             elif opdunits.lower() == 'nanometer' or opdunits.lower() == 'nm':
   2092                 self.opd *= 1e-9

ValueError: output array is read-only
```

### PR DESCRIPTION
**Pull request by josePhoenix**
Fixes #89.

### CODE REVIEW

**Comment by josePhoenix** [comment]
There's some noise in here because I didn't keep my `master` clean... rest assured, it's a one line change.

File: poppy/poppy_core.py [hunk 1]
```diff
@@ -2059,7 +2059,7 @@ def __init__(self, name="unnamed optic", transmission=None, opd= None, opdunits=
             elif isinstance(opd, fits.HDUList):
                 # load from fits HDUList
                 self.opd_file='supplied as fits.HDUList object'
-                self.opd = opd[0].data
+                self.opd = opd[0].data.copy()
                 self.opd_header = opd[0].header
```
Comment (mperrin) [inline poppy/poppy_core.py]: Should we also copy the FITS header, for consistency? I can't think of any practical circumstances where header edits mid calculation are likely to occur, but it seems odd to copy the data array but not the header.

Comment (josePhoenix) [inline poppy/poppy_core.py]: I can't either, but I guess it is more consistent. When implementing this, I noticed HDUList didn't have a `copy()` method and assumed `fits.Header` didn't either, but I was mistaken. I'll add it.

Comment (mperrin) [inline poppy/poppy_core.py]: I generally use `deepcopy` to copy entire HDUlists and all of their contents in one fell swoop, but that may be overkill in this case.

**Pull request merged**